### PR TITLE
ceph-disk: ability to use a different cluster name with dmcrypt

### PR DESCRIPTION
**Pull request by leseb**
Prior to this commit we were not able to configure an OSD using dmcrypt
on a cluster with a different name than 'ceph'. Adding the command line
option to --cluster <cluster name> to fix this.

Signed-off-by: Sébastien Han <seb@redhat.com>

### CODE REVIEW

File: src/ceph-disk/ceph_disk/main.py [hunk 1]
```diff
@@ -2377,13 +2379,14 @@ def set_or_create_partition(self):
                       self.args.lockbox)
             self.partition = self.create_partition()
 
-    def create_key(self):
+    def create_key(self, cluster):
```
Comment (alfredodeza) [inline src/ceph-disk/ceph_disk/main.py]: there is no need for requiring cluster here, you can rely on `self.args` since that will have the arguments passed in:

    self.args.cluster

File: src/ceph-disk/ceph_disk/main.py [hunk 2]
```diff
@@ -2377,13 +2379,14 @@ def set_or_create_partition(self):
...
         key_size = CryptHelpers.get_dmcrypt_keysize(self.args)
         key = open('/dev/urandom', 'rb').read(key_size / 8)
         base64_key = base64.b64encode(key)
         command_check_call(
             [
                 'ceph',
+                '--cluster', cluster,
```
Comment (alfredodeza) [inline src/ceph-disk/ceph_disk/main.py]: you would be able to just use that `self.args.cluster` in here, which already defaults to `ceph` if not explicitly defined

**Comment by ktdreyer** [comment]
Would you please file a ticket at http://tracker.ceph.com for this, and add the URL of the ticket to the commit log, so we can track this and ensure it is backported to Jewel eventually?

**Comment by leseb** [comment]
@ktdreyer sure, see: http://tracker.ceph.com/issues/17821

**Comment by leseb** [comment]
I'll update the PR with a definitive fix today.

**Comment by tchaikov** [comment]
@leseb could you put the line of `Fixes: http://tracker.ceph.com/issues/17821` into your commit message, right before your "SIgned-off-by" line, see https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst#tag-the-commit .

**Comment by leseb** [comment]
@tchaikov just did, thanks!
I still have some work to do on that patch.

**Comment by ktdreyer** [comment]
Please remove "DNM" from the title when this PR is ready for review and merging.

**Comment by leseb** [comment]
@ErwanAliasr1 and I reworked the patch, ready for review now :)

**Comment by ErwanAliasr1** [comment]
WDYT about this patch series guys ?

**Comment by ktdreyer** [comment]
Would you please rebase this onto the latest master to ensure that Jenkins is happy with it?

**Comment by leseb** [comment]
@ktdreyer done

**Comment by dachary** [comment]
this needs rebasing (and the test fails)

File: src/ceph-disk/ceph_disk/main.py [hunk 3]
```diff
@@ -4743,6 +4758,12 @@ def make_activate_parser(subparsers):
         action='store_true', default=False,
         help='activate the deactived OSD',
     )
+    activate_parser.add_argument(
+        '--cluster',
+        metavar='NAME',
+        default='ceph',
+        help='ceph cluster name (default: ceph)',
+    )
```
Comment (dachary) [inline src/ceph-disk/ceph_disk/main.py]: Activate does not have a --cluster argument because it can be determined with the find_cluster_by_uuid() function

Comment (leseb) [inline src/ceph-disk/ceph_disk/main.py]: Alright, let me try again without this, I forgot the error that made us think we needed to do this.

**Comment by leseb** [comment]
Ok testing is blocked by https://github.com/ceph/ceph/pull/12033/files waiting for the backport to re-validate.

**Comment by leseb** [comment]
@dachary I just pushed a new version, just removed the commit that brings --cluster support to activate since this is not needed.

**Comment by alfredodeza** [comment]
Jenkins test this please

**Comment by alfredodeza** [comment]
@leseb looks like you have some valid test failures:

    E       AssertionError: Expected call: get_dmcrypt_key('UUID', '/etc/ceph/dmcrypt-keys', True)
    E       Actual call: get_dmcrypt_key('UUID', '/etc/ceph/dmcrypt-keys', True, 'ceph')

**Comment by leseb** [comment]
@alfredodeza where can I get more details about these errors?

**Comment by alfredodeza** [comment]
@leseb the logs expired, we need to try this again

**Comment by alfredodeza** [comment]
jenkins test this please

**Comment by leseb** [comment]
@alfredodeza not sure if my patch is responsible for this error, @dachary do you mind having a quick look into this please? Thanks!

**Comment by alfredodeza** [comment]
@leseb these failures are related to the changes you've made. Specifically to the `get_dmcrypt_key()` function, which it has been altered to receive the extra argument of `'ceph'`.

That is why the test fails, because it was crafted to expect a strict set of arguments that didn't include the cluster name.

Since that is no longer the case, the test fails. You need to update the test to make sure that `'ceph'` is now expected.

The test failure occurs in: ../src/ceph-disk/tests/test_prepare.py line 319 

The actual failure is:

    E       AssertionError: Expected call: get_dmcrypt_key('UUID', '/etc/ceph/dmcrypt-keys', True)
    E       Actual call: get_dmcrypt_key('UUID', '/etc/ceph/dmcrypt-keys', True, 'ceph')

The assertion that needs to change is:

            assert m['_dmcrypt_map'].called
            m['get_dmcrypt_key'].assert_called_with(
               uuid, '/etc/ceph/dmcrypt-keys', True)

**Comment by leseb** [comment]
@alfredodeza thanks for the detailed explanation, just pushed a change that should hopefully fix this issue.

**Comment by leseb** [comment]
jenkins test this please

**Comment by leseb** [comment]
@alfredodeza all green now :)
Can someone merge this? Btw @dachary I addressed your comment so if you're ok with the patch please approve the PR :)
Thanks!

File: src/ceph-disk/ceph_disk/main.py [hunk 4]
```diff
@@ -2443,7 +2451,7 @@ def populate(self):
         LOG.debug('Mounting lockbox ' + str(" ".join(args)))
         command_check_call(args)
         write_one_line(path, 'osd-uuid', self.args.osd_uuid)
-        self.create_key()
+        self.create_key(self.args.cluster)
```
Comment (dachary) [inline src/ceph-disk/ceph_disk/main.py]: the self.create_key method should use self.args.cluster: there is no need to pass it as an argument

Comment (leseb) [inline src/ceph-disk/ceph_disk/main.py]: done

**Comment by dachary** [comment]
Did you run a ceph-disk suite (teuthology --suite ceph-disk) successfully with this pull request ?

**Comment by leseb** [comment]
@dachary I have not, can you help me with this?

**Comment by dachary** [comment]
@leseb just push your branch to ceph-ci and I'll run it for you.

**Comment by leseb** [comment]
@dachary like this? https://github.com/ceph/ceph-ci/pull/4

**Comment by neurodrone** [comment]
It'd be nice to have this as a part of 10.2.6.

**Comment by leseb** [comment]
@dachary can you run the CI with https://github.com/ceph/ceph-ci/pull/4?

**Comment by dachary** [comment]
@leseb you need to push a branch on ceph-ci for teuthology to be run

**Comment by leseb** [comment]
@dachary I don't have permission to push a branch to that repo, that's why I created a PR...

**Comment by dachary** [comment]
@leseb you have permission to push, all ceph developers have permission and you're in the group, it must be something on your side that's not configured right.

**Comment by leseb** [comment]
@dachary please see:

```
$ git remote -v |grep ceph-cii
ceph-cii        git@github.com:ceph/ceph-ci.git (push)
ceph-cii        git@github.com:ceph/ceph-ci.git (fetch)
$ git push ceph-cii dmcrypt-cluster-name
ERROR: Permission to ceph/ceph-ci.git denied to leseb.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

Am I doing something wrong?

**Comment by dachary** [comment]
I don't see what's wrong. Here are the permissions you have:

seb

I guess someone with better knowledge of github subtle settings can help out :-)

**Comment by leseb** [comment]
@dachary seems to be working now, branch pushed :)

**Comment by dachary** [comment]
@leseb nice :-) What was the problem ? Do you have access to the sepia lab ?

**Comment by leseb** [comment]
@dachary well looks like I wasn't part of Ceph team, then I got an email saying you added me and it was good :).

**Comment by dachary** [comment]
> well looks like I wasn't part of Ceph team, then I got an email saying you added me and it was good :).

Oo amazing. I was looking to verify if you were listed and mistook the "add" box for the "search" box... happy ending :-) Welcome to the Ceph team, always nice to have new members ...

**Comment by yuriw** [comment]
@leseb @dachary @GregMeno 
tests failed http://pulpito.front.sepia.ceph.com/yuriw-2017-02-15_21:45:11-ceph-disk-wip-yuri-testing2_2017_2_16-distro-basic-vps/

wip-yuri-testing2_2017_2_16

**Comment by GregMeno** [comment]
http://pulpito.ceph.com/yuriw-2017-02-15_21:45:11-ceph-disk-wip-yuri-testing2_2017_2_16-distro-basic-vps/

**Comment by GregMeno** [comment]
http://pulpito.ceph.com/yuriw-2017-02-15_21:45:11-ceph-disk-wip-yuri-testing2_2017_2_16-distro-basic-vps/

**Comment by neurodrone** [comment]
Will this be backported to Jewel by any chance?

**Comment by vumrao** [comment]
@neurodrone this is the tracker - http://tracker.ceph.com/issues/17821, it was marked backport to jewel and I have changed it to pending backport. It will get backported to jewel.

**Comment by vasukulkarni** [comment]
@dachary it would have been nice to default the cluster to 'ceph' here, we got a patch for ceph-deploy to add the cluster option there, Also would be nice to run ceph-deploy suite for ceph-disk changes as well.

**Comment by ganeshmaharaj** [comment]
@dachary @vasukulkarni Should have updated both threads. https://github.com/ceph/ceph/pull/13527 is the patch submitted to ceph-disk to add the default.

**Comment by dachary** [comment]
@leseb I was wrong merging this pull request. I misread https://github.com/ceph/ceph/pull/11786#issuecomment-280190074 and thought the tests were successfull. Instead they **failed** and as a result ceph-disk is now broken in master. Could you please fix this asap ?

**Comment by leseb** [comment]
@dachary I originally proposed the same approach as in https://github.com/ceph/ceph/pull/13527 so we should merge it.

**Comment by dachary** [comment]
@leseb of course, entirely my fault, please accept my apologies

**Comment by ganeshmaharaj** [comment]
@dachary Thanks for the update. i believe with this,https://github.com/ceph/ceph-deploy/pull/430 will be needed for ceph-deploy with work. @vasukulkarni

**Comment by leseb** [comment]
@dachary No worries :)

**Comment by GregMeno** [comment]
@leseb @dachary  Would you please help me understand the ceph-disk suite failure here?
cheers,
G

**Comment by dachary** [comment]
@GregMeno ceph-disk activate does not have a --cluster argument but the function implementing it expects it. It does not show in the ceph-disk suite output because the part that fails is run via either systemd or udev.
